README
Calendar-Hijri

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After  installing, you  can  find documentation for this module with  the perldoc
command.

    perldoc Calendar::Hijri

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Calendar-Hijri

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Calendar-Hijri

    CPAN Ratings
        http://cpanratings.perl.org/d/Calendar-Hijri

    Search CPAN
        http://search.cpan.org/dist/Calendar-Hijri/

LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2014 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Calendar::Hijri

0.34  Wed Nov 21 14:15:00 2018
      - Used namespace::autoclean instead.

0.33  Wed Jan 25 11:30:00 2017
      - Upgraded to use Date::Hijri::Simple v0.18.

0.32  Sun Jan 08 03:40:00 2017
      - Added dependency on Calendar::Plugin::Renderer v0.12.
      - Removed method validate_params(), no longer required.

0.31  Fri Dec 30 11:45:00 2016
      - Added sample code for method as_text() in the SYNOPSIS section.
      - Updated pod document for method as_text().

0.30  Mon Dec 19 12:50:00 2016
      - Upgraded to use Date::Hijri::Simple v0.17.

0.29  Sun Dec 11 07:40:00 2016
      - Added test scripts (t/meta-json.t and t/meta-yml.t).

0.28  Thu Nov 10 10:05:00 2016
      - Upgraded to use Date::Hijri::Simple v0.16.

0.27  Mon Oct 17 11:40:00 2016
      - Added dependency on Calendar::Plugin::Renderer v0.11.

0.26  Sat Sep 10 20:10:00 2016
      - Added key 'provides' to the Makefile.PL script.

0.25  Tue Aug 09 10:50:00 2016
      - Upgraded to use Date::Hijri::Simple v0.15.

0.24  Wed Jul 06 15:00:00 2016
      - Added method as_text().
      - Updated method as_svg() with regard to start index.
      - Added dependency on Calendar::Plugin::Renderer v0.07.
      - Added test for method as_svg().

0.23  Wed May 18 13:45:00 2016
      - Upgraded to use Date::Hijri::Simple v0.12.
      - Updated method as_svg() to handle month name.
      - Refactored method current() and method from_gregorian().

0.22  Sat May 07 13:10:00 2016
      - Added method as_svg() using the plugin Calendar::Plugin::Renderer.
      - Added section "SEE ALSO".
      - Updated copyright year information.

0.21  Sun Aug 30 11:30:00 2015
      - Upgraded to use Date::Hijri::Simple v0.09.

0.20  Fri Jul 17 12:40:00 2015
      - Upgraded to use Date::Hijri::Simple v0.08.
      - Making use of the method get_calendar() provided by Date::Hijri::Simple.

0.19  Wed Jul 08 09:05:00 2015
      - Upgraded to use Date::Hijri::Simple v0.07.

0.18  Wed Jul 01 10:20:00 2015
      - Fixed the pod document showing sample Hijri Calendar.

0.17  Sun Jun 28 02:05:00 2015
      - Upgraded the pre-requisite Date::Hijri::Simple v0.06.
      - Updated method _calendar() w.r.t the renamed method days_in_month_year().

0.16  Thu Jun 18 10:20:00 2015
      - Fixed the bug in the method from_gregorian().

0.15  Wed Jun 10 10:25:00 2015
      - Updated method _calendar() to validate month and year.

0.14  Mon May 25 03:05:00 2015
      - Removed the dependency on Date::Utils::Hijri.
      - Upgraded the pre-requisite Date::Hijri::Simple v0.03.
      - Fine tuned method BUILD().
      - Fine tuned method current().
      - Fine tuned method from_julian().
      - Updated method _calendar().
      - Added sample calendar to the section "SYNOPSIS".
      - Added sample code to the section "SYNOPSIS".
      - Updated pod description of method from_gregorian().
      - Removed sample code from pod document for method current().
      - Removed sample code from pod document for method from_gregorian().
      - Removed sample code from pod document for method from_julian().

0.13  Thu Apr 23 23:40:00 2015
      - Re-structured and consumed Date::Utils::Hijri as Moo Role.

0.12  Tue Apr 21 13:50:00 2015
      - Fixed the bug in the method _calendar() the last cell is empty.

0.11  Sun Apr 19 11:15:00 2015
      - Complete re-write using Moo with unit tests.

0.10  Mon Mar 16 13:30:10 2015
      - Added LICENSE file.
      - Tidied up pod document in general.

0.09  Fri Jan 16 14:55:10 2015
      - Removed SIGNATURE file.
      - Updated MANIFEST.SKIP file to ignore folder 'eumm/'.
      - Updated Copyright year information.

0.08  Sat Oct 18 22:30:19 2014
      - Converted all files into unix format.

0.07  Sun Sep 14 18:32:20 2014
      - Deleted Build.PL file.
      - Removed dependency of Readonly.pm and Carp.pm.
      - Added section "REPOSITORY" to the pod
      - Added key "META_MERGE" to Makefile.PL

0.06  Tue Aug 09 16:21:11 2011
      - Made Readonly v1.03 as minimum requirement.
      - Prepared the distribution using the Module::Signature v0.68.

0.05  Tue Apr 19 10:52:34 2011
      - Added description to the pod of method days_so_far().

0.04  Tue Apr 19 09:21:13 2011
      - Fixed the typo in the description.

0.03  Mon Apr 18 15:52:00 2011
      - Renamed title "SYNOPSIS" to "DESCRIPTION".
      - Renamed method days_in_year_month() to days_in_month().
      - Renamed methos days_up_until_year_month() to days_so_far().
      - Added example pod for method from_gregorian().
      - Added example pod for method to_gregorian().
      - Added example pod for method to_julian().
      - Cosmetics changes.

0.02  Tue Apr 12 12:37:37 2011
      - Fixed the internal method _validate_date().
      - Added unit test 01-test-case.t

0.01  Mon Apr 11 18:34:23 2011
      - First version released.


Makefile.PL
#!/usr/bin/perl

use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Calendar::Hijri',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Calendar/Hijri.pm',
    ABSTRACT_FROM      => 'lib/Calendar/Hijri.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More'        => 0,
        'File::Temp'        => 0,
        'XML::SemanticDiff' => 0,
    },
    PREREQ_PM => {
        'Moo'                        => '2.000000',
        'namespace::autoclean'       => '0.28',
        'Date::Hijri::Simple'        => '0.18',
        'Calendar::Plugin::Renderer' => '0.12',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Calendar-Hijri-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Calendar::Hijri' => { file => 'lib/Calendar/Hijri.pm', version => '0.34' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Calendar-Hijri.git',
                web  => 'https://github.com/manwar/Calendar-Hijri',
            },
        }})
     : ()
    ),
);


lib/Calendar/Hijri.pm
package Calendar::Hijri;

=head1 NAME

Calendar::Hijri - Interface to Islamic Calendar.

=head1 VERSION

Version 0.34

=cut

use 5.006;
use Data::Dumper;

use Date::Hijri::Simple;
use Moo;
use namespace::autoclean;
with 'Calendar::Plugin::Renderer';

use overload q{""} => 'as_string', fallback => 1;

$Calendar::Hijri::VERSION   = '0.34';
$Calendar::Hijri::AUTHORITY = 'cpan:MANWAR';

has year  => (is => 'rw', predicate => 1);
has month => (is => 'rw', predicate => 1);
has date  => (is => 'ro', default   => sub { Date::Hijri::Simple->new });

sub BUILD {
    my ($self) = @_;

    $self->date->validate_year($self->year)   if $self->has_year;
    $self->date->validate_month($self->month) if $self->has_month;

    unless ($self->has_year && $self->has_month) {
        $self->year($self->date->year);
        $self->month($self->date->month);
    }
}

=head1 DESCRIPTION

Hijri Calendar begins with the migration from Mecca to Medina of Mohammad (pbuh),
the Prophet of Islam, an event  known  as the Hegira. The initials A.H.  before a
date mean "anno Hegirae" or "after Hegira". The  first  day  of the year is fixed
in the Quran as the first day of the month of Muharram.In 17 AH Umar I,the second
caliph, established the beginning of the era of the Hegira ( 1 Muharram 1 AH ) as
the date that is 16 July 622 CE in the Julian Calendar.

The years are lunar & consist of 12 lunar months. There is no intercalary period,
since the Quran ( Sura IX, verses 36,37 )  sets  the calendar year  at 12 months.
Because the year in the Hijri  calendar is shorter than a solar year, the  months
drift with respect to the seasons, in a cycle 32.50 years.

    +--------------------------------------------------------------------------------------------------------+
    |                                       Sha'aban        [1436 BE]                                        |
    +--------------+--------------+--------------+--------------+--------------+--------------+--------------+
    |      al-Ahad |   al-Ithnayn | ath-Thulatha |     al-Arbia |    al-Khamis |    al-Jumuah |      as-Sabt |
    +--------------+--------------+--------------+--------------+--------------+--------------+--------------+
    |                                            |            1 |            2 |            3 |            4 |
    +--------------+--------------+--------------+--------------+--------------+--------------+--------------+
    |            5 |            6 |            7 |            8 |            9 |           10 |           11 |
    +--------------+--------------+--------------+--------------+--------------+--------------+--------------+
    |           12 |           13 |           14 |           15 |           16 |           17 |           18 |
    +--------------+--------------+--------------+--------------+--------------+--------------+--------------+
    |           19 |           20 |           21 |           22 |           23 |           24 |           25 |
    +--------------+--------------+--------------+--------------+--------------+--------------+--------------+
    |           26 |           27 |           28 |           29 |                                            |
    +--------------+--------------+--------------+--------------+--------------+--------------+--------------+

The package L<App::calendr> provides command line tool  C<calendr> to display the
supported calendars on the terminal.

=head1 SYNOPSIS

    use strict; use warnings;
    use Calendar::Hijri;

    # prints current hijri month calendar.
    print Calendar::Hijri->new, "\n";
    print Calendar::Hijri->new->current, "\n";

    # prints hijri month calendar for the first month of year 1436.
    print Calendar::Hijri->new({ month => 1, year => 1436 }), "\n";

    # prints hijri month calendar in which the given gregorian date falls in.
    print Calendar::Hijri->new->from_gregorian(2015, 1, 14), "\n";

    # prints hijri month calendar in which the given julian date falls in.
    print Calendar::Hijri->new->from_julian(2457102.5), "\n";

    # prints current month hijri calendar in SVG format.
    print Calendar::Hijri->new->as_svg;

    # prints current month hijri calendar in text format.
    print Calendar::Hijri->new->as_text;

=head1 HIJRI MONTHS

    +--------+------------------------------------------------------------------+
    | Number | Name                                                             |
    +--------+------------------------------------------------------------------+
    |   1    | Muharram                                                         |
    |   2    | Safar                                                            |
    |   3    | Rabi' al-awwal                                                   |
    |   4    | Rabi' al-thani                                                   |
    |   5    | Jumada al-awwal                                                  |
    |   6    | Jumada al-thani                                                  |
    |   7    | Rajab                                                            |
    |   8    | Sha'aban                                                         |
    |   9    | Ramadan                                                          |
    |  10    | Shawwal                                                          |
    |  11    | Dhu al-Qi'dah                                                    |
    |  12    | Dhu al-Hijjah                                                    |
    +--------+------------------------------------------------------------------+

=head1 HIJRI DAYS

    +--------------+------------------------------------------------------------+
    | Arabic Name  | English Name                                               |
    +--------------+------------------------------------------------------------+
    |      al-Ahad | Sunday                                                     |
    |   al-Ithnayn | Monday                                                     |
    | ath-Thulatha | Tuesday                                                    |
    |     al-Arbia | Wednesday                                                  |
    |    al-Khamis | Thursday                                                   |
    |    al-Jumuah | Friday                                                     |
    |      as-Sabt | Saturday                                                   |
    +--------------+------------------------------------------------------------+

=head1 METHODS

=head2 current()

Returns current month of the Hijri calendar.

=cut

sub current {
    my ($self) = @_;

    return $self->as_text($self->date->month, $self->date->year);
}

=head2 from_gregorian()

Returns Hijri month calendar in which the given gregorian date falls in.

=cut

sub from_gregorian {
    my ($self, $year, $month, $day) = @_;

    return $self->from_julian($self->date->gregorian_to_julian($year, $month, $day));
}

=head2 from_julian($julian_date)

Returns Hijri month calendar in which the given julian date falls in.

=cut

sub from_julian {
    my ($self, $julian) = @_;

    my $date = $self->date->from_julian($julian);
    return $self->as_text($date->month, $date->year);
}

=head2 as_text($month, $year)

Returns  color  coded  Hijri  calendar  for  the given C<$month> and C<$year>. If
C<$month> and C<$year> missing, it would return current calendar month.

=cut

sub as_text {
    my ($self, $month, $year) = @_;

    ($month, $year) = $self->validate_params($month, $year);
    my $date = Date::Hijri::Simple->new({ year => $year, month => $month, day => 1 });

    return $self->text_calendar(
        {
            start_index => $date->day_of_week,
            month_name  => $date->get_month_name,
            days        => $date->days_in_month_year($month, $year),
            year        => $year,
            day_names   => $date->days,
        });
}

=head2 as_svg($month, $year)

Returns calendar for the given C<$month> and C<$year> rendered  in SVG format. If
C<$month> and C<$year> missing, it would return current calendar month.

=cut

sub as_svg {
    my ($self, $month, $year) = @_;

    ($month, $year) = $self->validate_params($month, $year);
    my $date = Date::Hijri::Simple->new({ year => $year, month => $month, day => 1 });

    return $self->svg_calendar({
        start_index => $date->day_of_week,
        month_name  => $date->get_month_name,
        days        => $date->days_in_month_year($month, $year),
        year        => $year });
}

sub as_string {
    my ($self) = @_;

    return $self->as_text($self->month, $self->year);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Calendar-Hijri>

=head1 SEE ALSO

=over 4

=item L<Calendar::Bahai>

=item L<Calendar::Gregorian>

=item L<Calendar::Persian>

=item L<Calendar::Saka>

=back

=head1 BUGS

Please report any bugs / feature requests to C<bug-calendar-hijri at rt.cpan.org>
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Calendar-Hijri>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Calendar::Hijri

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Calendar-Hijri>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Calendar-Hijri>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Calendar-Hijri>

=item * Search CPAN

L<http://search.cpan.org/dist/Calendar-Hijri/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain  a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Calendar::Hijri
